### PR TITLE
revise SSR+hydration functionality to be more fault-tolerant

### LINKED ISSUE #3017: Fix Rehydration regex causing infinite loops for malformed inputs
The regex based rehydration can fall into an infinite loop in some edge cases. As part of this issue we should:

- Update the regex to prevent the edge case in the first place
- Investigate why it caused an infinite loop and protect against similar loops to prevent this issue in the future

Edit: I'm now actually not convinced it's an infinite loop. I've got a minimal reproduction that eventually completes after a very long time, so it may be a Regex catastrophic backtracking issue.

This issue was discovered in #2972. There some CSS was being output (malformed by stylis due to `/*!` comments) and caused the following rule to be input into rehydration:

```css
{/*! normalize.css v8.0.1 | MIT License | github.com/necolas/normalize.css *//*!                                                                                                                               
 *  Font Awesome 4.7.0 by @davegandy - http://fontawesome.io - @fontawesome                                                                                                                                    
 *  License - http://fontawesome.io/license (Font: SIL OFL 1.1, CSS: MIT License)                                                                                                                              
 */} 
```

This causes an infinite loop since (presumably) it's a selector-less rule body.

Edit: Because I forgot as well, to recap why we switched away from CSSOM-based rehydration in #2872: There are quirks and bugs in browsers that cause `cssText` on rules to be unreliable 🤦‍♂ See: https://bugzilla.mozilla.org/show_bug.cgi?id=1594241

Edit 2: This test here gets slower as more characters (one by one!) are added to the invalid `normalize.css` comment line:

```js
  it.only('tolerates malformed CSS', () => {
    document.head.innerHTML = `
      <style ${SC_ATTR} ${SC_ATTR_VERSION}="${SC_VERSION}">
        {/*! normalize.css v8.
          }
        .rule {}
        ${SC_ATTR}.g1[id="idA"]{content:""}
      </style>
    `;

    const sheet = new StyleSheet({ isServer: true });
    rehydrateSheet(sheet);
  });
```

Edit 3: Content doesn't matter so it seems to somehow be the newline in the rule, which usually is stripped out by stylis, so this may be why it's such a rare edge case (?)

Edit 4: Using `/([^{]*){((?:{[^}]*}|(?!{).*?)*)}/g;` instead (removing `.*`) and adding `trim` later helps to keep the time down but after a couple more characters the time grows again. So there's another backtracking issue in here.

### PR DESCRIPTION
**Pull request by kitten**
Fix #3017

### CODE REVIEW

File: packages/styled-components/src/sheet/Rehydration.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ import { getIdForGroup, setGroupForId } from './GroupIDAllocator';
 import type { Sheet } from './types';
 
 const SELECTOR = `style[${SC_ATTR}][${SC_ATTR_VERSION}="${SC_VERSION}"]`;
-const RULE_RE = /(?:\s*)?(.*?){((?:{[^}]*}|(?!{).*?)*)}/g;
+const RULE_RE = /([^{]*){((?:[^{}]+(?:{[^}]*})?)*)}/g;
```
Comment (kitten) [inline packages/styled-components/src/sheet/Rehydration.js]: - Leading `(?:\s*)?` is gone. This isn't needed because `exec` will keep walking forwards character by character. Also we want to start matching as soon as possible and if we later don't match, then this will slow the regex down because it'll start backtracking.
- Replaced `.*?` (btw `*?` is redundant) with `[^{]*` because we know that a `{` must follow. So we safe some time here by telling the Regex that we're matching as much as possible until that char instead of eagerly matching when the following characters don't quite match.
- Internally in the regex we had the same problems but more pronounced.

Comment (probablyup) [reply packages/styled-components/src/sheet/Rehydration.js]: This doesn't work for certain types of deeply nested media queries unfortunately, use this as a test string:

```
@media (max-width:300px){.fvsNoc{display:block;}@supports (display:-webkit-box) or (display:-webkit-flex) or (display:-ms-flexbox) or (display:flex){.fvsNoc{display:-webkit-box;display:-webkit-flex;display:-ms-flexbox;display:flex;}}@media (max-width:320px){.fvsNoc{display:block;}}}
data-styled.g1[id="pages__Container-sc-4wa63o-0"]{content:"fvsNoc,"}
.dBiKfw{font-size:8rem;font-weight:900;color:darkgreen;text-align:center;}
data-styled.g2[id="pages__Heading-sc-4wa63o-1"]{content:"dBiKfw,"}
```

(it's from the original issue I created the regex from)

The revised regex breaks at the second `@media`

Comment (probablyup) [reply packages/styled-components/src/sheet/Rehydration.js]: My suggestion here is to change the SSR emitter to add a special character CSS doesn't process to serve as the "split" point (for use in `string.split()` and just ditch the splitting regex.

Comment (kitten) [reply packages/styled-components/src/sheet/Rehydration.js]: We could likely just add `/*!sc*/` as splitting markers

File: packages/styled-components/src/sheet/Rehydration.js [hunk 2]
```diff
@@ -53,15 +52,12 @@ const rehydrateNamesFromContent = (sheet: Sheet, id: string, content: string) =>
...
-
-  // parts = [match, selector, content]
-  // eslint-disable-next-line no-cond-assign
-  while ((parts = RULE_RE.exec(rawHTML))) {
-    const marker = parts[1].match(MARKER_RE);
 
+  for (let i = 0, l = parts.length; i < l; i++) {
+    const part = parts[i].trim();
+    const marker = part.match(MARKER_RE);
     if (marker) {
```
Comment (probablyup) [inline packages/styled-components/src/sheet/Rehydration.js]: Let's add a `continue` shortcircuit at the beginning if `part` is empty. Since we're using the `split` method it's possible there might be crud in there

**Comment by probablyup** [comment]
Published for testing under `styled-components@5.0.1-revisedssr`

File: packages/styled-components/src/sheet/test/GroupedTag.test.js [hunk 3]
```diff
@@ -89,7 +89,7 @@ it('does supports large group numbers', () => {
   expect(groupedTag.length).toBeGreaterThan(group);
   expect(tag.length).toBe(1);
   expect(groupedTag.indexOfGroup(group)).toBe(0);
-  expect(groupedTag.getGroup(group)).toBe('.test {}\n');
+  expect(groupedTag.getGroup(group)).toBe('.test {}/*!sc*/\n');
```
Comment (abbyhowell) [inline packages/styled-components/src/sheet/test/GroupedTag.test.js]: I don't understand why this change adds this string to the output of this method. How do we know this change isn't introducing a new, unwanted behavior, if all these tests needed to change?

File: packages/styled-components/src/sheet/test/Rehydration.test.js [hunk 4]
```diff
@@ -94,4 +100,19 @@ describe('rehydrateSheet', () => {
...
   });
+
+  it('tolerates long, malformed CSS', () => {
+    document.head.innerHTML = `
+      <style ${SC_ATTR} ${SC_ATTR_VERSION}="${SC_VERSION}">
+        {xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+          xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
+          }
+        .rule {}/*!sc*/
+        ${SC_ATTR}.g1[id="idA"]{content:""}/*!sc*/
```
Comment (abbyhowell) [inline packages/styled-components/src/sheet/test/Rehydration.test.js]: Maybe this test CSS should include a real inline comment with `/*!` wrapped around it, rather than only `sc`? What am I missing here?

File: packages/styled-components/src/sheet/test/Rehydration.test.js [hunk 5]
```diff
@@ -23,31 +23,37 @@ describe('outputSheet', () => {
...
-      '.b {}',
-      `${SC_ATTR}.g22[id="idB"]{content:"nameB,"}`,
-    ]);
+      .split('/*!sc*/');
+
+    expect(output).toMatchInlineSnapshot(`
+      Array [
+        ".a {}",
+        "
+      data-styled.g11[id=\\"idA\\"]{content:\\"nameA,\\"}",
```
Comment (abbyhowell) [inline packages/styled-components/src/sheet/test/Rehydration.test.js]: `data-styled` should probably still be `${SC_ATTR}` here, to accommodate the different `process.env` values it could be (https://github.com/styled-components/styled-components/blob/d0e8eab8cf391630294c6ac9e81f206eb64e5e53/packages/styled-components/src/constants.js#L6:14)

File: packages/styled-components/src/sheet/test/Rehydration.test.js [hunk 6]
```diff
@@ -23,31 +23,37 @@ describe('outputSheet', () => {
...
+        "
+      .b {}",
+        "
+      data-styled.g22[id=\\"idB\\"]{content:\\"nameB,\\"}",
```
Comment (abbyhowell) [inline packages/styled-components/src/sheet/test/Rehydration.test.js]: Why is the output double-escaped now when it wasn't before? That seems like an unwanted change.

Comment (probablyup) [reply packages/styled-components/src/sheet/test/Rehydration.test.js]: I think this is just a jest snapshot serialization artifact

File: packages/styled-components/src/sheet/test/Rehydration.test.js (hunk 4 above)
Comment (probablyup) [reply packages/styled-components/src/sheet/test/Rehydration.test.js]: This is just the new "splitter" marker we are adding. You can consider it an implementation detail for SSR.

File: packages/styled-components/src/sheet/test/GroupedTag.test.js (hunk 3 above)
Comment (probablyup) [reply packages/styled-components/src/sheet/test/GroupedTag.test.js]: This is used by `outputSheet` in the SSR functionality. We need to add the marker at this point so the individual rules are split correctly during rehydration. This PR generally changes the whole rehydration mechanic, which previously used regexes to do the splitting.
